core/src/main/scala/BinaryTree.scala

```diff
@@ -0,0 +1,250 @@
+/**
+ * Created by Nicolas A Perez (@anicolaspp) on 1/29/16.
+ */
+
+package dogs
+
+import Predef._
+import dogs.Order.{GT, EQ, LT}
+import scala.annotation.tailrec
+import scala.math
+
+
+
+abstract class BinaryTree[A] {
+
+  def value: Option[A]
+
+  val left: BinaryTree[A]
+  val right: BinaryTree[A]
+
+  private [dogs] val size: Int
+  private[dogs] val height: Int
+  private [dogs] def balanceFactor: Int
+  private[dogs] def update(): BinaryTree[A]
+
+  def isEmpty: Boolean
+
+  def toLst(): List[A] = this match {
+    case BTNil()                      =>  El[A]
+    case Branch(Some(a), l, r, _, _)  =>  l.toLst ::: (a :: r.toLst)
```

Comment:
the fact that you never use these last two parameters (height and size) in any pattern match ever should be a sign that they don't need to be part of the constructor.